internal/pkg/build/assemblers/sif.go

```diff
@@ -189,38 +190,61 @@ func (a *SIFAssembler) Assemble(b *types.Bundle, path string) error {
 	}
 	sylog.Verbosef("Set SIF container architecture to %s", arch)
 
-	if err := s.Create([]string{b.RootfsPath}, fsPath, flags); err != nil {
-		return fmt.Errorf("while creating squashfs: %v", err)
-	}
-
 	var encOpts *encryptionOptions
-
-	if b.Opts.EncryptionKeyInfo != nil {
-		plaintext, err := cryptkey.NewPlaintextKey(*b.Opts.EncryptionKeyInfo)
-		if err != nil {
-			return fmt.Errorf("unable to obtain encryption key: %+v", err)
+	if b.Opts.Unprivilege {
+		sylog.Debugf("Creating squashfs image and will use gocryptfs")
+		if b.Opts.EncryptionKeyInfo == nil {
+			sylog.Errorf("no encryption key environment variable or --passphrase provided")
+			return fmt.Errorf("no encryption key environment variable or --passphrase provided")
 		}
 
-		// A dm-crypt device needs to be created with squashfs
-		cryptDev := &crypt.Device{}
+		g := packer.NewGocryptfs(b.Opts.EncryptionKeyInfo)
+		g.MksquashfsPath = a.MksquashfsPath
 
-		// TODO (schebro): Fix #3876
-		// Detach the following code from the squashfs creation. SIF can be
-		// created first and encrypted after. This gives the flexibility to
-		// encrypt an existing SIF
-		loopPath, err := cryptDev.EncryptFilesystem(fsPath, plaintext)
-		if err != nil {
-			return fmt.Errorf("unable to encrypt filesystem at %s: %+v", fsPath, err)
+		if err := g.Create([]string{b.RootfsPath}, fsPath, flags, b.TmpDir); err != nil {
+			return fmt.Errorf("while employing gocryptfs to create image, err: %v", err)
 		}
-		defer os.Remove(loopPath)
-
-		fsPath = loopPath
 
 		encOpts = &encryptionOptions{
 			keyInfo:   *b.Opts.EncryptionKeyInfo,
-			plaintext: plaintext,
+			plaintext: []byte(g.Pass),
+		}
+	} else {
+		sylog.Debugf("Creating squashfs img and might use device-mapper if encryption is requested")
```

Comment:
I think this would be better as two debug messages, one here saying "Creating squashfs image" and another one a little below when it is using device-mapper encryption just saying "Using device-mapper encryption".

Reply:
resolved in this commit
https://github.com/apptainer/apptainer/pull/1185/commits/17c45e4d5d0eab05651c9877307fcb7d58282232